File: packages/cli/src/__tests__/auth.test.ts
import { describe, it, expect } from "vitest";
import {
  buildAuthFlowState,
  applyDeviceCode,
  applyApiKeyInput,
  type AuthProvider,
} from "../lib/auth.js";

const copilot: AuthProvider = {
  id: "copilot",
  name: "GitHub Copilot",
  method: "device_code",
  ready: false,
};
const anthropic: AuthProvider = {
  id: "anthropic",
  name: "Anthropic",
  method: "api_key",
  envVar: "ANTHROPIC_API_KEY",
  ready: true,
};
const openai: AuthProvider = {
  id: "openai",
  name: "OpenAI",
  method: "api_key",
  envVar: "OPENAI_API_KEY",
  ready: false,
};

describe("buildAuthFlowState", () => {
  it("filters out ready providers", () => {
    const flow = buildAuthFlowState([copilot, anthropic, openai]);
    expect(flow.providers).toHaveLength(2);
    expect(flow.providers.map((p) => p.id)).toEqual(["copilot", "openai"]);
  });

  it("returns empty providers when all ready", () => {
    const flow = buildAuthFlowState([anthropic]);
    expect(flow.providers).toHaveLength(0);
  });

  it("returns all when none ready", () => {
    const flow = buildAuthFlowState([copilot, openai]);
    expect(flow.providers).toHaveLength(2);
  });
});

describe("applyDeviceCode", () => {
  it("sets deviceCode on existing flow", () => {
    const flow = buildAuthFlowState([copilot]);
    const result = applyDeviceCode(flow, {
      userCode: "ABCD-1234",
      verificationUri: "https://github.com/login/device",
    });
    expect(result?.deviceCode?.userCode).toBe("ABCD-1234");
    expect(result?.deviceCode?.verificationUri).toBe("https://github.com/login/device");
  });

  it("returns null when flow is null", () => {
    expect(applyDeviceCode(null, { userCode: "X", verificationUri: "Y" })).toBeNull();
  });

  it("preserves existing providers", () => {
    const flow = buildAuthFlowState([copilot, openai]);
    const result = applyDeviceCode(flow, { userCode: "X", verificationUri: "Y" });
    expect(result?.providers).toHaveLength(2);
  });
});

describe("applyApiKeyInput", () => {
  it("sets apiKeyInput with provider info", () => {
    const flow = buildAuthFlowState([copilot, openai]);
    const result = applyApiKeyInput(flow, "openai", "OpenAI");
    expect(result?.apiKeyInput?.providerId).toBe("openai");
    expect(result?.apiKeyInput?.providerName).toBe("OpenAI");
  });

  it("looks up provider name from flow when not provided", () => {
    const flow = buildAuthFlowState([copilot, openai]);
    const result = applyApiKeyInput(flow, "openai");
    expect(result?.apiKeyInput?.providerName).toBe("OpenAI");
  });

  it("uses providerId as fallback name", () => {
    const flow = buildAuthFlowState([copilot]);
    const result = applyApiKeyInput(flow, "unknown");
    expect(result?.apiKeyInput?.providerName).toBe("unknown");
  });

  it("returns null when flow is null", () => {
    expect(applyApiKeyInput(null, "openai")).toBeNull();
  });
});

File: test/opal/auth_test.exs
defmodule Opal.AuthTest do
  use ExUnit.Case, async: true

  alias Opal.Auth

  describe "probe/0" do
    test "returns a map with status, provider, and providers" do
      result = Auth.probe()

      assert Map.has_key?(result, :status)
      assert Map.has_key?(result, :provider)
      assert Map.has_key?(result, :providers)
      assert result.status in ["ready", "setup_required"]
      assert is_list(result.providers)
      assert length(result.providers) == 4
    end

    test "all providers include id, name, method, and ready" do
      result = Auth.probe()

      for provider <- result.providers do
        assert Map.has_key?(provider, :id)
        assert Map.has_key?(provider, :name)
        assert Map.has_key?(provider, :method)
        assert Map.has_key?(provider, :ready)
        assert is_boolean(provider.ready)
      end
    end

    test "providers include copilot, anthropic, openai, google" do
      result = Auth.probe()
      ids = Enum.map(result.providers, & &1.id)
      assert "copilot" in ids
      assert "anthropic" in ids
      assert "openai" in ids
      assert "google" in ids
    end
  end

  describe "ready?/1" do
    test "returns boolean for copilot" do
      assert is_boolean(Auth.ready?("copilot"))
    end

    test "returns boolean for anthropic" do
      assert is_boolean(Auth.ready?("anthropic"))
    end

    test "returns false for unknown provider" do
      refute Auth.ready?("unknown_provider")
    end
  end
end
